### PR TITLE
fix: Swagger file upload

### LINKED ISSUE #2628: Bug: File upload example doesn't work with Swagger
### Summary

The examples demonstrating file uploads in the docs don't work well with Swagger.

[Example from docs](https://docs.litestar.dev/2/usage/requests.html#file-uploads)
```py
from typing import Annotated

from litestar import Litestar, MediaType, post
from litestar.datastructures import UploadFile
from litestar.enums import RequestEncodingType
from litestar.params import Body


@post(path="/", media_type=MediaType.TEXT)
async def handle_file_upload(
    data: Annotated[UploadFile, Body(media_type=RequestEncodingType.MULTI_PART)],
) -> str:
    content = await data.read()
    filename = data.filename
    return f"{filename}, {content.decode()}"


app = Litestar(route_handlers=[handle_file_upload])
```

# Result
http://127.0.0.1:8000/schema/swagger#/default/HandleFileUpload
image

# Expected
image

https://swagger.io/docs/specification/2-0/file-upload/

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh Litestar dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2628">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2628/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by guacs**
### Pull Request Checklist

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [ ] Pre-Commit Checks were ran and passed
- [ ] Tests were ran and passed

### Description

- Adds the value for `format` as `binary` as per the OpenAPI 3.0 spec.
- For multiple file upload, the [spec](https://swagger.io/docs/specification/describing-request-body/file-upload/) requires the type of the request body schema to be `object` with `properties` instead of a schema of type `array` and `items`.

Before:
image

After:
image

### Close Issue(s)

- Fixes #2628

### CODE REVIEW

**Review by guacs — COMMENTED** [review]
I missed checking for a list of files and the fix made here does not seem to work. I'll have to investigate that a bit more.

**Comment by provinzkraut** [comment]
Any updates on this @guacs?

**Comment by guacs** [comment]
> Any updates on this @guacs?

I got stuck on this due to the multiple file upload bug, but since that's been fixed, I think I'll be able to get this ready today.

**Review by cofin — APPROVED** [review]
Thanks for taking care of this one @guacs

**Comment by peterschutt** [comment]
I can't get the single file upload to work:

```py
from typing import List

from typing_extensions import Annotated

from litestar import Litestar, MediaType, post
from litestar.datastructures import UploadFile
from litestar.enums import RequestEncodingType
from litestar.params import Body


@post(path="/", media_type=MediaType.TEXT)
async def handle_file_upload(
    data: Annotated[UploadFile, Body(media_type=RequestEncodingType.MULTI_PART)],
) -> str:
    content = await data.read()
    filename = data.filename
    return f"{filename}, {content.decode()}"


@post(path="/s", media_type=MediaType.TEXT)
async def handle_file_uploads(data: Annotated[List[UploadFile], Body(media_type=RequestEncodingType.MULTI_PART)]) -> str:
    contents = [(await d.read()).decode() for d in data]
    filenames = [d.filename for d in data]
    return f"{filenames}, {contents}"


app = Litestar(route_handlers=[handle_file_upload, handle_file_uploads], debug=True)
```
Whenever I upload to the single file route:

```json
{"status_code":400,"detail":"Validation failed for POST http://127.0.0.1:8000/","extra":[{"message":"Unsupported type: <class 'dict'>","key":"data","source":"body"}]}
```

Uploading the same file via the multi-file route works though.

For the single file route the request payload is empty except for a single boundary:

```
-----------------------------249385238218377964211058925597--
```

While the multi file one:

```
-----------------------------16023743603440668548827953628
Content-Disposition: form-data; name="file"; filename="gitignore"
Content-Type: application/octet-stream

.venv/
.idea/
.vscode/
.python-version
.tox/
.pre-commit-config.yaml

-----------------------------16023743603440668548827953628--
```

**Review by guacs — COMMENTED** [review]
Test the single file upload again.

EDIT: @peterschutt can you add a review comment on this? I don't wanna accidentally end up merging this. Apparently it doesn't work if I add the review comment.

**Comment by guacs** [comment]
> This fix definitely gets the upload UI to display properly, however I can't get the single file upload to work:
> 
> ```py
> from typing import List
> 
> from typing_extensions import Annotated
> 
> from litestar import Litestar, MediaType, post
> from litestar.datastructures import UploadFile
> from litestar.enums import RequestEncodingType
> from litestar.params import Body
> 
> 
> @post(path="/", media_type=MediaType.TEXT)
> async def handle_file_upload(
>     data: Annotated[UploadFile, Body(media_type=RequestEncodingType.MULTI_PART)],
> ) -> str:
>     content = await data.read()
>     filename = data.filename
>     return f"{filename}, {content.decode()}"
> 
> 
> @post(path="/s", media_type=MediaType.TEXT)
> async def handle_file_uploads(data: Annotated[List[UploadFile], Body(media_type=RequestEncodingType.MULTI_PART)]) -> str:
>     contents = [(await d.read()).decode() for d in data]
>     filenames = [d.filename for d in data]
>     return f"{filenames}, {contents}"
> 
> 
> app = Litestar(route_handlers=[handle_file_upload, handle_file_uploads], debug=True)
> ```
> Whenever I upload to the single file route:
> 
> ```json
> {"status_code":400,"detail":"Validation failed for POST http://127.0.0.1:8000/","extra":[{"message":"Unsupported type: <class 'dict'>","key":"data","source":"body"}]}
> ```
> 
> Uploading the same file via the multi-file route works though.
> 
> For the single file route the request payload is empty except for a single boundary:
> 
> ```
> -----------------------------249385238218377964211058925597--
> ```
> 
> While the multi file one:
> 
> ```
> -----------------------------16023743603440668548827953628
> Content-Disposition: form-data; name="file"; filename="gitignore"
> Content-Type: application/octet-stream
> 
> .venv/
> .idea/
> .vscode/
> .python-version
> .tox/
> .pre-commit-config.yaml
> 
> -----------------------------16023743603440668548827953628--
> ```
> 

That's weird. Let me check it out and see what happened.

**Review by peterschutt — DISMISSED** [review]
See comment in PR description.

**Comment by peterschutt** [comment]
> Test the single file upload again.
> 
> EDIT: @peterschutt can you add a review comment on this? I don't wanna accidentally end up merging this. Apparently it doesn't work if I add the review comment.

Yep, done!

File: litestar/_openapi/schema_generation/schema.py
Comment (guacs) [inline litestar/_openapi/schema_generation/schema.py:387]: I'm using `field_definition.kwarg_definition is None` as a proxy to figure out whether this `FieldDefinition` was created directly from the `data` field or whether it was a part of another class. 

So, for the following `field_definition.kwarg_definition` would not be `None`.

```python
@post("/")
async def handle_file_upload(
    data: Annotated[UploadFile, Body(media_type=RequestEncodingType.MULTI_PART)]) -> None:
    ....
```

For the `FieldDefinitions` that reach `for_upload_file` in the below example, the `kwarg_definition` would be `None`.
```python
class Foo(Struct):
    foo: UploadFile
    bar: UploadFile

@post("/f")
async def foo_file(data: Annotated[Foo, Body(media_type=RequestEncodingType.MULTI_PART)]) -> None:
```

The reason this differentiation is needed is because, the populating of `properties` of the schema for `Foo` will be handled by the logic of `for_msgspec_struct` etc. However, if the type is just `UploadFile` like in the first example, then the `properties` needs to be populated within `for_upload_file` itself.

@peterschutt are there any problems with this logic?

**Comment by guacs** [comment]
@peterschutt can you test this out and see if it's all working?

**Comment by guacs** [comment]
Also, I haven't updated the tests yet (which is why they're failing). Once it's confirmed that it's working, I'll update the tests etc.

**Comment by JacobCoffee** [comment]
Ive got a few internal apps that use file upload, ill test this

**Comment by guacs** [comment]
> Ive got a few internal apps that use file upload, ill test this

@JacobCoffee did you get a chance to test this?

**Comment by JacobCoffee** [comment]
no ._. I will do this shortly

**Comment by JacobCoffee** [comment]
Ahh right I remember why I couldnt - I am having an issue with some kind of database connection that is required for my app to run... I'm going to have to keep troubleshooting but don't hold up the progress/mering if it works in testing for you guys. Hopefully I can get back on this tomorrow because I'm angry at it tonight 😓

**Comment by guacs** [comment]
> Ahh right I remember why I couldnt - I am having an issue with some kind of database connection that is required for my app to run... I'm going to have to keep troubleshooting but don't hold up the progress/mering if it works in testing for you guys. Hopefully I can get back on this tomorrow because I'm angry at it tonight 😓

lol I feel ya. I was so annoyed at the OpenAPI spec yesterday when trying to get this to work :P

**Comment by JacobCoffee** [comment]
got my shit fixed, testing now

**Comment by JacobCoffee** [comment]
not sure I have done this correctly, but this is what i get:
`pip install git+https://github.com/litestar-org/litestar.git@swagger-file-upload` 
image

**Comment by JacobCoffee** [comment]
Worked w/ peters example app in multi and single. So probably just a me thing... this is what I am doing for uploads:

```py
async def upload_files(
        self,
        data: FileUpload = Body(media_type=RequestEncodingType.MULTI_PART),
    ) -> list[str]:
       print("yep")
    
    class FileUpload(PydanticBaseModel):
	    files:list[UploadFile] = Field(description="Files to upload.")
	
	    # noinspection PyNestedDecorators
	    @field_validator("files", mode="before")
	    @classmethod
	    def parse_files(cls, value: UploadFile | list[UploadFile]) -> list[UploadFile]:
	
	        if isinstance(value, list):
	            return value
	        if isinstance(value, UploadFile):
	            return [value]
	        raise ValueError(value)
  ```

**Comment by guacs** [comment]
> Worked w/ peters example app in multi and single. So probably just a me thing... this is what I am doing for uploads:
> 
> ```python
> from litestar import Controller, Litestar, post
> from litestar.datastructures import UploadFile
> from litestar.enums import RequestEncodingType
> from litestar.params import Body
> from pydantic import BaseModel, Field, field_validator
> 
> 
> class PydanticBaseModel(BaseModel):
> 
>     class Config:
>         arbitrary_types_allowed = True
> 
> class FileUpload(PydanticBaseModel):
>     files: list[UploadFile] = Field(description="Files to upload.")
> 
>     # noinspection PyNestedDecorators
>     @field_validator("files", mode="before")
>     @classmethod
>     def parse_files(cls, value: UploadFile | list[UploadFile]) -> list[UploadFile]:
>         if isinstance(value, list):
>             return value
>         if isinstance(value, UploadFile):
>             return [value]
>         raise ValueError(value)
> 
> 
> class FileController(Controller):
>     @post(path="upload")
>     async def upload_files(self, data: FileUpload = Body(media_type=RequestEncodingType.MULTI_PART)) -> list[str]:
>         return [file.filename for file in data.files]
> 
> 
> app = Litestar(route_handlers=[FileController])
> ```

It doesn't look like you're doing anything wrong. I'll look into this, and update.

**Comment by guacs** [comment]
@JacobCoffee, @peterschutt I think I finally got it to work now. These are the different scenarios I tested:

```python
from __future__ import annotations

from pydantic import BaseModel

from litestar import Controller, Litestar, get, post
from litestar.datastructures import UploadFile
from litestar.enums import RequestEncodingType
from litestar.params import Body


class PydanticBaseModel(BaseModel):
    class Config:
        arbitrary_types_allowed = True


class FileUpload(PydanticBaseModel):
    files: list[UploadFile]
    file: UploadFile


class FileController(Controller):
    @post("/single")
    async def upload_single_file(self, data: UploadFile = Body(media_type=RequestEncodingType.MULTI_PART)) -> str:
        return data.filename

    @post("/list")
    async def upload_files_list(
        self, data: list[UploadFile] = Body(media_type=RequestEncodingType.MULTI_PART)
    ) -> list[str]:
        return [f.filename for f in data]

    @post(path="/object")
    async def upload_files_object(
        self, data: FileUpload = Body(media_type=RequestEncodingType.MULTI_PART)
    ) -> dict[str, list[str] | str]:
        return {"files": [f.filename for f in data.files], "file": data.file.filename}


@get("/favicon.ico")
async def favicon() -> None:
    return None


app = Litestar(route_handlers=[FileController, favicon], debug=True)
```

All of these were working when I tested it so hopefully it's all good now. If one of y'all could just confirm that things are working as expected (except the scenario I mention below), then that'd be great :)

NOTE: There is one scenario where it doesn't work and that's if you give only a single file to `FileUpload.files` which expects a list of files. This is a bug in the extraction logic we have and is related to #2753. I'll raise an issue for that with a minimal reproduction.

File: litestar/_openapi/schema_generation/schema.py
Comment (guacs) [reply litestar/_openapi/schema_generation/schema.py:366]: I think the correct solution is to just give return the same schema that we would if it was `list[UploadFile]` since that's what we're really expecting from the client. It's just that we convert that into a dictionary internally for ease of use. Any thoughts on this?

**Comment by JacobCoffee** [comment]
I will test this tomorrow at work!

**Comment by vkuznetsovgn** [comment]
I just tested this branch with my project and everything worked perfectly! Hopefully, we will see it in the release soon :)

**Comment by guacs** [comment]
> I just tested this branch with my project and everything worked perfectly! Hopefully, we will see it in the release soon :)

Oh that's fantastic! @JacobCoffee one you've tested, it'd be great if you could give it a review :)

**Comment by JacobCoffee** [comment]
Please excuse my delay, but it looked good to me when I tested :)

image

**Review by JacobCoffee — APPROVED** [review]

**Comment by guacs** [comment]
@peterschutt I dismissed your review because I couldn't find a way to mark it as resolved (even though it has been resolved) and without that I can't merge :P

**Comment by peterschutt** [comment]
> @peterschutt I dismissed your review because I couldn't find a way to mark it as resolved (even though it has been resolved) and without that I can't merge :P

No problem!

**Pull request merged**
